guidance/models/_model.py

```diff
@@ -209,645 +72,113 @@ def get_chat_template(self): # TODO [HN]: Add more logic here...should we instan
-                    response_is_generated,
-                    response_new_bytes_prob,
-                    response_capture_groups,
-                    response_capture_group_log_probs,
-                    response_new_token_count,
-                ) = response_state
-
-                yield EngineCallResponse(
-                    new_bytes=response_new_bytes,
-                    is_generated=response_is_generated,
-                    new_bytes_prob=response_new_bytes_prob,
-                    capture_groups=response_capture_groups,
-                    capture_group_log_probs=response_capture_group_log_probs,
-                    new_token_count=response_new_token_count,
-                )
-
-            if logits_state is not None:
-                token_ids, forced_bytes, current_temp = logits_state
-                logits = self.get_logits(token_ids, forced_bytes, current_temp)
-
-            if is_done:
-                break
-
-    def _tokenize_prefix(self, byte_string):
-        """This is used to speed up the tokenization of long prompts without using the parser."""
-        token_ids = []
-        token_byte_positions = []
-
-        # loop trying to decode a new token at each iteration
-        pos = 0
-        while True:
-
-            # walk down the token trie looking for a unique token match
-            trie = self._token_trie
-            valid_pos = -1
-            valid_value = -1
-            while True:
-                if pos >= len(byte_string):
-                    if len(trie) > 0:
-                        valid_pos = -1
-                    break
-
-                # check if we can keep going or are at a dead end
-                if trie.has_child(byte_string[pos : pos + 1]):
-                    trie = trie.child(byte_string[pos : pos + 1])
-                    pos += 1
-
-                    # record the last valid token down this path as we go
-                    if trie.value >= 0:
-                        valid_pos = pos
-                        valid_value = trie.value
+        parser = self.start(prompt, grammar, ensure_bos_token)
+
+        token = None
+        while not parser.done():
+            gen_data, response = parser.advance(token)
+
+            if gen_data is not None:
+                if parser.is_accepting() and self.tokenizer.eos_token_id is not None:
+                    # Whenever we are in an accepting state, we will allow the model to generate whatever it wants
+                    # but we will treat any "illegal" tokens as EOS, allowing the model to finish gracefully.
+                    assert gen_data.mask[self.tokenizer.eos_token_id]
+                    token = self.get_next_token(
+                        token_ids=gen_data.tokens,
+                        mask=None,
+                        temperature=gen_data.temperature
+                    )
+                    if not gen_data.mask[token]:
+                        token = self.tokenizer.eos_token_id
```

Comment:
@Harsha-Nori would appreciate an eye on this. 

FYI: `is_accepting()` just means "if we send an EOS, the grammar will be done", e.g. if we have the grammar`gen(regex=r"\d+")` and produce the string "12345" -- we can choose to stop now or keep going.

I believe that this condition is equivalent to what is in existing guidance.

*Alternatively*, we could relax this condition from `is_accepting` to `bool(gen_data.mask[self.tokenizer.eox_token_id])`. There are a few places where we allow EOS that aren't necessarily the end of the full grammar, e.g. as the implicit stop condition of a `gen` call with no passed `stop` arg... Any opinions on this? I know it's kind of subtle...